pkg/kubeapiserver/authenticator/config.go

```diff
@@ -310,6 +316,23 @@ func (c *authenticationConfigUpdater) updateAuthenticationConfig(ctx context.Con
 		return utilerrors.NewAggregate([]error{lastErr, waitErr}) // filters out nil errors
 	}
 
+	// Reset the metrics for the jwt authenticators, since we have a new set of them
+	// and don't want the metrics to be stale or grow unbounded.
+	// This will reset the following metrics:
+	// - jwtAuthenticatorLatencyMetric
+	// - jwksFetchLastTimestampSeconds
+	// - jwksFetchLastKeySetHash
+	// as they are split by the jwt issuer hash.
+	//
+	// We swap the new authenticator in the next step, and cancel the old ones after a minute.
+	// This can lead to a brief period when the old authenticators are still in use and might
+	// record metrics. These stale metrics would get reset in the next reload cycle and the
+	// metrics will not grow unbounded.
+	//
+	// TODO(aramase): Investigate if we want to optimize this behavior in the v1.32 release
+	// to prevent the brief period of stale metrics.
+	oidc.ResetMetrics()
```

Comment:
is it a problem to clear metrics for a jwt authenticator that was previously in use and is still in use?

Reply:
The JWKS fetch metric from old authenticator for same issuer will not be relevant for the new authenticator. As part of the first `AuthenticateToken` request, the new authenticator will _need_ to fetch the keyset that'll be success/failure. The old metrics for the same issuer will not provide any info about the current state of the new authenticator after authentication config reloading.

Reply:
It is certainly not great, hence the TODO to look into making the metrics less global and more per authenticator.  I figured a global reset was safer than risking any kind of memory leak if someone was cycling through authenticators.

Reply:
(also Anish is correct in regards to semantics of reload+metrics for same authenticator per issuer being rather nuanced and non-obvious)

Reply:
so just to be clear, if we reset them here, the old ones will still be reporting in the background for another ~minute
* if the names overlap, old / new will duel for a minute and the metrics won't make sense until the churn is done?
* if the names don't overlap, the old ones will pop back into published metrics until the *next* reload?

Reply:
> if we reset them here, the old ones will still be reporting in the background for another ~minute

The most likely metric that'll be generated for the old ones is the latency metric. For the JWKS fetch ones because the old authenticator already has it, chances are low it'll re-fetch and update the metric (but it's still possible)

> if the names overlap, old / new will duel for a minute and the metrics won't make sense until the churn is done?

yes, both old/new will duel and update the metrics for that brief period. For the JWKS fetch metrics, the chances of this value being the same are high for the same issuer url (diff might only be if they used the same issuer but different discovery url with diff key set in the new authn config).

 > if the names don't overlap, the old ones will pop back into published metrics until the next reload?

this is correct! During that minute, if the old authenticators are validating token, they will update the latency metric. They _probably_ will not update the JWKS fetch metrics because the authenticator already has the keyset cached. These stale metrics will go away after the next reload.

Reply:
just a thought: if we used the authentication config hash as a metric label, we could cleanly reset all the metrics of authenticators in old authn config _after_ the swap

Reply:
> I figured a global reset was safer than risking any kind of memory leak if someone was cycling through authenticators.

is metric growth only a risk if issuers are changing? given the negatives of a global reset, should we only do that if issuers are changing between the old / new config?